crates/solver/src/liquidity/zeroex.rs

```diff
@@ -79,36 +127,177 @@ impl ZeroExLiquidity {
+/// messages.
+async fn connect_socket() -> Result<Socket> {
+    let url = Url::parse("wss://api.0x.org/orderbook/v1")?;
+    let mut socket = tokio_tungstenite::connect_async(url).await?.0;
+
+    // Construct the subscription message to fetch all orders
+    let subscription_msg = serde_json::json!({
+        "type": "subscribe",
+        "channel": "orders",
+        "requestId": "cowswap",
+    });
+
+    socket
+        .send(Message::Text(subscription_msg.to_string()))
+        .await?;
+
+    Ok(socket)
+}
+
+/// Creates a websocket connection and reads the messages from it.
+/// The messages are then used to update the cache.
+/// If the connection is lost, calling again this function will reconnect.
+async fn connect_and_update_cache(
+    cache: Arc<Mutex<HashMap<OrderId, OrderRecord>>>,
+) -> Result<(), UpdateError> {
+    let mut socket = connect_socket().await?;
+    let result = update_cache(&mut socket, cache.clone()).await;
+    // this call will error if the socket is already closed but that's fine
+    let _ = socket.close(None).await;
+    result
+}
+
+#[derive(Debug)]
+enum UpdateError {
+    // The websocket is closed on server side, with optional reason
+    SocketClosed(Option<anyhow::Error>),
+    // Received a message type that we don't handle
+    UnsupportedMessage,
+    // Received proper `text` type of message but it can't be deserialized (mailformed or changed
+    // format)
+    DeserializeError(anyhow::Error),
+    Other(anyhow::Error),
+}
+
+impl From<anyhow::Error> for UpdateError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::Other(err)
+    }
+}
+
+impl From<UpdateError> for anyhow::Error {
+    fn from(err: UpdateError) -> Self {
+        match err {
+            UpdateError::SocketClosed(err) => anyhow::anyhow!("socket closed, reason {:?}", err),
+            UpdateError::UnsupportedMessage => anyhow::anyhow!("unsupported message"),
+            UpdateError::DeserializeError(err) => err,
+            UpdateError::Other(err) => err,
+        }
+    }
+}
+
+async fn update_cache(
+    socket: &mut Socket,
+    cache: Arc<Mutex<HashMap<OrderId, OrderRecord>>>,
+) -> Result<(), UpdateError> {
+    while let Some(msg) = socket.next().await {
+        let msg = msg.context("websocket error")?;
+        let text = match msg {
+            Message::Text(text) => text,
+            Message::Close(frame) => {
+                return Err(UpdateError::SocketClosed(
+                    frame.map(|frame| anyhow!(frame.reason)),
+                ))
+            }
+            Message::Ping(payload) => {
+                // send pong message to keep the connection alive
+                socket
+                    .send(Message::Pong(payload))
+                    .await
+                    .context("ping pong failure")?;
+                continue;
+            }
+            _ => {
+                tracing::debug!("Received unsupported message {:?}", msg);
```

Comment:
Should this be logged as an error so we fix unsupported messages?

Reply:
Agree.